src/internal/internal_hettmqr.cc

```diff
@@ -278,48 +304,32 @@ void hettmqr(
                         tile::deepConjTranspose(C(j, i1));
                     }
                 }
-                else if (C.tileIsLocal(i2, j)) {
-                    // Sends updated tile back.
-                    int src = (i1 >= j
-                              ? C.tileRank(i1, j)
-                              : C.tileRank(j, i1));
-                    // Send updated tile back.
-                    C.tileSend(i1, j, src, tag);
-                    if (call_tile_tick) {
-                        C.tileTick(i1, j);
-                    }
-                }
             } // for j
-        } // for index
-
-        //--------------------
-        // Finish updating all rows before updating columns.
-        slate_mpi_call(
-            MPI_Barrier(C.mpiComm()));
+        }
+        #pragma omp taskwait
 
+        requests.clear();
         for (int index = 0; index + step < nranks; index += 2*step) {
             int64_t j1 = rank_indices[ index ].second;
             int64_t j2 = rank_indices[ index + step ].second;
 
             //--------------------
             // 3: Multiply [ C(i, j1)  C(i, j2) ] * Q for i = j2+1, ..., mt-1.
             for (int64_t i = j2+1; i < C.mt(); ++i) {
+                int tag = tag_base + i + j1*C.mt();
                 if (C.tileIsLocal(i, j1)) {
                     // First node of each pair sends tile to dst.
                     int dst = C.tileRank(i, j2);
-                    C.tileSend(i, j1, dst, tag);
+                    MPI_Request req;
+                    C.tileIsend( i, j1, dst, tag, &req );
+                    MPI_Request_free( &req );
```

Comment:
Why do you free the `MPI_Request` here?

Reply:
This process won't touch tile `(i, j1)` until this process receives the tile back, which won't resolve before the tile is send from the remote process, which in turn won't happen until the remote process receives the tile from this process.  And there's enough MPI calls in this sections that forward progress shouldn't be an issue.